fenicsprecice/fenicsprecice.py

```diff
@@ -183,42 +201,31 @@ def read_data(self):
-                vertices = np.zeros((n_vertices, 2))
-                vertices[:, 0] = self._coupling_mesh_vertices[:, 0]
-                vertices[:, 1] = self._coupling_mesh_vertices[:, 1]
-                # z is the dead direction so the data is not transferred to vertices
-                assert (np.sum(np.abs(self._coupling_mesh_vertices[:, 2])) < 1e-10)
-            else:
-                raise Exception("Dimensions do not match.")
-        elif self._read_function_type is FunctionType.VECTOR:
-            if self._fenics_dimensions == self._interface.get_dimensions():
+        read_data = None
+
+        if self._empty_rank:
+            assert (self._size > 1)  # having participants without coupling mesh nodes is only valid for parallel runs
+
+        if not self._empty_rank:
+            if self._read_function_type is FunctionType.SCALAR:
+                read_data = self._interface.read_block_scalar_data(read_data_id, self._vertex_ids)
+            elif self._read_function_type is FunctionType.VECTOR:
                 read_data = self._interface.read_block_vector_data(read_data_id, self._vertex_ids)
-                vertices = self._coupling_mesh_vertices
-            elif self._apply_2d_3d_coupling:
-                precice_read_data = self._interface.read_block_vector_data(read_data_id, self._vertex_ids)
-                n_vertices, dims = precice_read_data.shape
-                read_data = np.zeros((n_vertices, dims - 1))
-                read_data[:, 0] = precice_read_data[:, 0]
-                read_data[:, 1] = precice_read_data[:, 1]
-
-                vertices = np.zeros((n_vertices, dims - 1))
-                vertices[:, 0] = self._coupling_mesh_vertices[:, 0]
-                vertices[:, 1] = self._coupling_mesh_vertices[:, 1]
-                # z is the dead direction so the data is not transferred to read_data array and vertices
-                assert (np.sum(np.abs(precice_read_data[:, 2])) < 1e-10)
-                assert (np.sum(np.abs(self._coupling_mesh_vertices[:, 2])) < 1e-10)
-            else:
-                raise Exception("Dimensions do not match.")
-        else:
-            raise Exception("Rank of function space is neither 0 nor 1")
 
-        return {tuple(key): value for key, value in zip(vertices, read_data)}
+            owned_read_data = {tuple(key): value for key, value in zip(self._owned_coords, read_data)}
+            # print("Rank {}: Data before communication: {}".format(self._rank, owned_read_data))
+            updated_data = communicate_shared_vertices(self._comm, self._rank, self._fenics_gids,
+                                                       self._owned_coords, self._fenics_coords, owned_read_data,
+                                                       self._to_send_pts, self._to_recv_pts)
+            # print("Rank {}: Data after communication: {}".format(self._rank, updated_data))
+        else:  # if there are no vertices, we return empty data
+            updated_data = None
```

Comment:
```suggestion
            read_data = {tuple(key): value for key, value in zip(self._owned_coords, read_data)}
            read_data = communicate_shared_vertices(self._comm, self._rank, self._fenics_gids,
                                                       self._owned_coords, self._fenics_coords, read_data,
                                                       self._to_send_pts, self._to_recv_pts)
        else:  # if there are no vertices, we return empty data
            read_data = None
```
See comment above. I think appending to `read_data` instead of redefining it inside `communicate_shared_vertices` is the cleaner solution.